src/renderers/shared/fiber/ReactFiberClassComponent.js

```diff
@@ -107,6 +177,7 @@ module.exports = function(scheduleUpdate : (fiber: Fiber, priorityLevel : Priori
     const ctor = workInProgress.type;
     const props = workInProgress.pendingProps;
     const instance = new ctor(props);
+    checkClassInstance(workInProgress, instance);
```

Comment:
Actually, this is not sufficient since the state may depend on `props` so it can yield different results during the rerender. This would then only throw an error during race conditions.

```js
class Foo extends React.Component {
  state = this.props.bar ? {} : [];
}
```

I'd rather check it every time.

For warnings, we should probably ideally dedupe though.